site/src/xServices/workspaces/workspacesXService.ts

```diff
@@ -206,160 +206,255 @@ interface WorkspacesContext {
+                  entry: "clearGetWorkspacesError",
+                  invoke: {
+                    src: "getWorkspaces",
+                    id: "getWorkspaces",
+                    onDone: [
+                      {
+                        target: "done",
+                        cond: "isEmpty",
+                        actions: "assignWorkspaceRefs",
+                      },
+                      {
+                        target: "updatingWorkspaceRefs",
+                      },
+                    ],
+                    onError: [
+                      {
+                        target: "done",
+                        actions: "assignGetWorkspacesError",
+                      },
+                    ],
+                  },
+                },
+                done: {
+                  type: "final",
+                },
+              },
+            },
+          },
+          onDone: {
+            target: "waitToRefreshWorkspaces",
           },
         },
       },
     },
-  },
-  {
-    guards: {
-      isEmpty: (context) => !context.workspaceRefs,
-    },
-    actions: {
-      assignWorkspaceRefs: assign({
-        workspaceRefs: (_, event) =>
-          event.data.map((data) => {
-            return spawn(workspaceItemMachine.withContext({ data }), data.id)
-          }),
-      }),
-      assignFilter: assign({
-        filter: (context, event) => event.query ?? context.filter,
-      }),
-      assignGetWorkspacesError: assign({
-        getWorkspacesError: (_, event) => event.data,
-      }),
-      clearGetWorkspacesError: (context) =>
-        assign({ ...context, getWorkspacesError: undefined }),
-      triggerUpdateVersion: (context, event) => {
-        const workspaceRef = context.workspaceRefs?.find(
-          (ref) => ref.id === event.workspaceId,
-        )
-
-        if (!workspaceRef) {
-          throw new Error(`No workspace ref found for ${event.workspaceId}.`)
-        }
-
-        workspaceRef.send("UPDATE_VERSION")
+    {
+      guards: {
+        isEmpty: (context) => !context.workspaceRefs,
       },
-      assignUpdatedWorkspaceRefs: assign({
-        workspaceRefs: (_, event) => {
-          const newWorkspaceRefs = event.data.newWorkspaces.map((workspace) =>
-            spawn(
-              workspaceItemMachine.withContext({ data: workspace }),
-              workspace.id,
-            ),
+      actions: {
+        assignWorkspaceRefs: assign({
+          workspaceRefs: (_, event) =>
+            event.data.map((data) => {
+              return spawn(workspaceItemMachine.withContext({ data }), data.id)
+            }),
+        }),
+        assignFilter: assign({
+          filter: (context, event) => event.query ?? context.filter,
+        }),
+        assignGetWorkspacesError: assign({
+          getWorkspacesError: (_, event) => event.data,
+        }),
+        clearGetWorkspacesError: (context) =>
+          assign({ ...context, getWorkspacesError: undefined }),
```

Comment:
curious why we overwrite the the whole context here instead of just updating the `getWorkspaceError` to be undefined.

Reply:
Oh so something not great about this diff is that I used the XState editor and it rearranged some code, so some stuff looks new that isn't. This is just an old pattern that we found a better way of writing since then.